feat: add `std::mem`

AlicanC opened the pull request:
This adds a new library to `sway-lib-std` called `mem`.

It implements:
- `alloc`, `copy`, `cmp`, `realloc`
- `struct Pointer`
- `struct Buffer` for allocating blocks of memory and working with them

nfurfaro reviewed (commented):
I think this looks good. I'd like to do some refactoring of other stuff to use this code!

otrho reviewed (changes requested):
This is great though I have a few thoughts:
* Eventually I'd like a lot of this to be replaced with intrinsics, though that's true for most of the ASM blocks in the stdlib.
* I'm a bit concerned about converting any old reference type to a pointer, and then assuming you can do what you like with it.  Especially if you take the address of a literal which is still a reference type, but resides in the data section, it is essentially read only memory, and so if you try to `copy()` to it, the VM will panic.  Maybe this isn't possible though, not sure.
* If we were to get a little more serious we could store the size of allocations in the heap.  This is what 'real' allocators do.  It allows `realloc()` to work without needing pass the old size in.  It also gives us `free()` although then we'd need to keep track of allocated and unallocated blocks for it to be useful.  But it also would blur the distinction between pointer and buffer in this code -- buffer wouldn't necessarily be needed, and we could use `Vec` when it arrives for collections of values.

AlicanC commented on sway-lib-std/src/mem.sw:
```diff
@@ -0,0 +1,209 @@
+/// Allocates an amount of memory on the heap
+pub fn alloc(size: u64) -> Pointer {
+    ~Pointer::new(asm(size: size, ptr) {
+        aloc size;
+        addi ptr hp i1;
+        ptr: u64
```
`$hp` points to the free space on the heap and not to the allocated space. Heap grows from end to start so you need `hp + 1` (and not `hp - size`) to get to the alloc'd space.

otrho replied:
Are you sure?  My impression from the spec is that the stack grows up, and so `sp` points to free space before you allocate.  Therefore you copy `sp` before `CFEI`.  Conversely, the heap grows down, so `hp` points to allocated memory before you allocate.  You copy `hp` after you `ALOC` and it'll just point to your new buffer.

So I think the above should just be `move ptr hp;`.  Adding a `1` really doesn't make sense either way.  If `hp` was 0x1000 and you allocate 256 bytes with `move r0 256; aloc r0` then `hp` will be 0xf00.  Adding 1 to it makes it 0xf01, which is just discards one of the bytes you allocated and actually makes the buffer 1 byte smaller.

Maybe @adlerjohn can confirm or elaborate?

AlicanC replied:
I wrote how I understood it in a comment:
```rs
/// Allocates zeroed memory on the heap
/// 
/// In FuelVM, the heap begins at `VM_MAX_RAM - 1` and grows downward.
/// Heap pointer `$hp` will always point to unallocated space.
/// 
/// Initially the heap will look like this:
/// ... 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |
///                                               $hp^  ^VM_MAX_RAM
/// 
/// After allocating with `let ptr = alloc(8)`:
/// ... 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |
///                       $hp^  ^ptr                    ^VM_MAX_RAM
/// 
/// After writing with `sw(ptr, u64::max())`:
/// ... 00 00 00 00 00 00 00 00 FF FF FF FF FF FF FF FF |
///                       $hp^  ^ptr                    ^VM_MAX_RAM
///
/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#vm-initialization
/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#aloc-allocate-memory
pub fn alloc(size: u64) -> u64 {
    asm(size: size, ptr) {
        aloc size;
        // `$hp` points to unallocated space and heap grows downward so
        // our newly allocated space will be right after it
        addi ptr hp i1;
        ptr: u64
    }
}
```

AlicanC commented on sway-lib-std/src/mem.sw:
```diff
@@ -0,0 +1,209 @@
+}
+
+impl Buffer {
+    /// Allocates a new buffer with the given size
+    pub fn new() -> Self {
+        Buffer {
+            ptr: alloc(0),
+            len: 0
+        }
+    }
```
The use is when you need a buf before starting to iterate or such:
```rs
let buf = ~Buffer::new();
while test {
  // resize and write to buf here
}
buf
```

`std::vec::RawVec` also does this:
https://github.com/FuelLabs/sway/pull/1118/files#diff-b5d9688741fea479477f26ca44cd1d1ecbd2f003f3875292abb23df7fad85c58R12-R19

otrho replied:
OK, this is kinda less than ideal.  It'd be better if the buffer could just be created after we know the size, but I guess sometimes that isn't possible.

AlicanC commented on sway-lib-std/src/mem.sw:
```diff
@@ -0,0 +1,209 @@
+        let ptr = if is_reference_type::<T>() {
+            ~Pointer::from(val)
+        } else {
+            ~Pointer::from((val))
```
My intention was to make compiler do the aloc and sw for me, but I'm actually going to make this revert when given a non-ref value just like `Pointer::from`. Going with "`from`s shouldn't allocate" for now.

adlerjohn replied:
As @otrho brings up, we'll have to consider how the optimizer will handle single-primitive structs, or potentially tuples. For now this is fine, but definitely warrants a sit-down session sometime.

AlicanC commented on sway-lib-std/src/mem.sw:
```diff
@@ -0,0 +1,209 @@
+            let ptr = ~Pointer::from(val);
+            let len = size_of::<T>();
+            
+            // We can't reference `self.write_unchecked` like this:
+            // self.write_unchecked(ptr, len, offset);
```
Will refer to it in a comment 👍

AlicanC commented:
@otrho thanks for the review!

My intention is to create an unsafe low level library to help w/ debugging and for building safe variants on later. `std::mem` is basically a collection of what I've been copying&pasting around for debugging script_data.

I've read the comments and made some research on what Rust has available and now I'm thinking:

- Add `std::alloc` a la Rust's [std::alloc](https://doc.rust-lang.org/std/alloc/index.html):
  - Add `std::alloc::Global`:
    - `alloc` = `alloc_zeroed`
    - `dealloc` = noop
  - Move alloc fns from `std::mem` here and map them to `std::alloc::Global`.
- Move non-alloc fns from `std::mem` to `std::intrinsic`:
  - I see that Rust has [std::intrinsics::copy](https://doc.rust-lang.org/std/intrinsics/fn.copy.html) and such so I guess that's suitable.
- Add `std::ptr` a la Rust's [std::ptr](https://doc.rust-lang.org/std/ptr/index.html):
  - Move `Pointer` here, but as `RawPointer` so we can later have `Pointer<T>`.
  - Add `size` to it, and move appropriate size-based methods from `Buffer` here.
- Remove `std::mem`:
  - At this point there won't be anything left in `std::mem` but a lighter `Buffer`. I will still need it because `Vec<u8>` will pad its `u8`s to 8 bytes so it's not really a "vector of bytes", but it will be easy to implement `Buffer` where I need to use it since `Pointer` will be doing the boring work.

Comments appreciated!

otrho commented:
Yep, cool, looks like a good plan!  I think we'll be trying to remove the `u8` is actually a `u64` problem down the track but not in the short term.

otrho commented on sway-lib-std/src/alloc.sw:
```diff
@@ -0,0 +1,45 @@
+
+use ::intrinsics::*;
+
+/// Allocates zeroed memory on the heap
+/// 
+/// In FuelVM, the heap begins at `VM_MAX_RAM - 1` and grows downward.
+/// Heap pointer `$hp` will always point to unallocated space.
+/// 
+/// Initially the heap will look like this:
+/// ... 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |
+///                                               $hp^  ^VM_MAX_RAM
+/// 
+/// After allocating with `let ptr = alloc(8)`:
+/// ... 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 |
+///                       $hp^  ^ptr                    ^VM_MAX_RAM
+/// 
+/// After writing with `sw(ptr, u64::max())`:
+/// ... 00 00 00 00 00 00 00 00 FF FF FF FF FF FF FF FF |
+///                       $hp^  ^ptr                    ^VM_MAX_RAM
+///
+/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#vm-initialization
+/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#aloc-allocate-memory
```
OK, I believe the initialisation of the VM is broken, and `$hp` should be initialised to `VM_MAX_RAM`.  Putting in hacks to account for the neglected last byte isn't the role of the stdlib.

Especially since if another part of the library used `ALOC` and didn't do the little `$hp + 1` trick then there'd be conflict and potential corruption.

otrho commented on sway-lib-std/src/alloc.sw:
```diff
@@ -0,0 +1,45 @@
+/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#vm-initialization
+/// See: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#aloc-allocate-memory
+pub fn alloc(size: u64) -> u64 {
+    asm(size: size, ptr) {
+        aloc size;
+        // `$hp` points to unallocated space and heap grows downward so
+        // our newly allocated space will be right after it
+        addi ptr hp i1;
+        ptr: u64
+    }
```
```suggestion
    asm(size: size) {
        aloc size;
        // `$hp` points to the new buffer.
        hp: u64
    }
```

AlicanC commented:
To elaborate on the `alloc` situation:

Fully implementing [Allocator](https://doc.rust-lang.org/stable/std/alloc/trait.Allocator.html) would look like this:
```rs
trait Allocator {
    fn allocate(self, size: u64) -> u64;
    fn deallocate(self, ptr: u64, size: u64) -> u64;
    fn allocate_zeroed(self, size: u64) -> u64;
    fn grow(self, ptr: u64, old_size: u64, new_size: u64) -> u64;
    fn grow_zeroed(self, ptr: u64, old_size: u64, new_size: u64) -> u64;
    fn shrink(self, ptr: u64, old_size: u64, new_size: u64) -> u64;
}
```
But in reality we only have these:
```rs
trait Allocator {
    fn allocate_zeroed(self, size: u64) -> u64;
    fn grow_zeroed(self, ptr: u64, old_size: u64, new_size: u64) -> u64;
}
```
And in my latest commit we call them `alloc` and `realloc`.

I see three options:
- Keep it as is: `alloc` and `realloc`
- Rename them to `allocate_zeroed`, `grow_zeroed`
- Implement full Allocator and start using deallocate/shrink/etc wherever makes sense even though they are noop.

otrho commented:
At this stage we really don't need to go apeshit with the `Allocator`, IMO, this is a great start and will do for now.  I'm happy except I still strongly feel the `$hp + 1` hack needs to go.

otrho commented:
There's no reason to make the + 1 fix now and merge.  We don't have to wait for the VM fix.  It'd just mean that until then the last byte in the heap will be ignored/unused.

adlerjohn commented:
Hmmm. But the VM's memory is `0` to `VM_MAX_RAM - 1` inclusive. I.e. `VM_MAX_RAM` isn't actually addressable.

otrho commented:
`$hp` starts at 0x3ffffff.  You allocate 0x10 bytes and now `$hp` is 0x3ffffef.  Your 0x10 byte buffer is all there.  No need to add 1 or anything since as far as I was aware we don't have alignment constraints.

After the VM fix `$hp` will start as 0x4000000, that's the only difference.  What am I missing?

adlerjohn commented:
Ah I got it, I was getting the direction mixed up. I though it was that the top byte of allocated mem would be outside the VM's memory, but it's instead that the top byte of the VM's memory wouldn't be allocatable. Right? In that case, that's fine.

AlicanC commented:
> There's no reason not to make the + 1 fix now and merge. We don't have to wait for the VM fix. It'd just mean that until then the last byte in the heap will be ignored/unused.

I need to return `$hp + 1` because you can't write to `$hp`.

otrho commented:
If you commit my suggestion above for `alloc()` and remove the comment, then it should be all fine and we can merge.

AlicanC commented:
I commited the suggestion, removed the comment and also removed `+1`s from the test and it's failing here: https://github.com/FuelLabs/sway/blob/ddd14809881456bde701261735f9110329b52a8b/test/src/sdk-harness/test_projects/alloc/src/main.sw#L37

```
  Compiled script "std_alloc_test".
  Bytecode size is 348 bytes.
    Finished test [unoptimized + debuginfo] target(s) in 0.20s
     Running test_projects/harness.rs (/Users/johncub/Repos/FuelLabs/sway/test/src/sdk-harness/target/debug/deps/integration_tests-b203013b97979b2e)

running 1 test
test alloc::alloc_implementation ... FAILED

failures:

---- alloc::alloc_implementation stdout ----
thread 'alloc::alloc_implementation' panicked at 'called `Result::unwrap()` on an `Err` value: ContractCallError("MemoryOverflow")', /Users/johncub/Repos/FuelLabs/sway/test/src/sdk-harness/test_helpers/src/lib.rs:31:47
```

This CI run should fail: https://github.com/FuelLabs/sway/runs/6605214966

otrho commented:
Alright, FYI, @AlicanC and I hopped in a huddle and we're seeing weirdness regarding writing to the allocated buffer.  I'm going to take a closer look at the VM implementation and see what's going on.

otrho commented:
An update:  I've made https://github.com/FuelLabs/fuel-specs/issues/322 and https://github.com/FuelLabs/fuel-vm/issues/137 because I believe the way we're using `$hp` and `ALOC` is not quite right.

But at the moment if we want to go ahead with this PR, we can use the original behaviour of expecting `$hp` to point to last byte in unallocated heap and using the `$hp + 1` idiom for newly allocated buffers.  We can switch it up if/when those issues are addressed.

AlicanC commented:
It seems like some change broke the way I use generics. Will try to isolate and open an issue.

AlicanC commented:
Resolved in https://github.com/FuelLabs/sway/pull/1979
